Structure From Motion

Drew-Wagner opened the pull request:
Created Structure From Motion Directory and added my files

Commit:
Initial Commit

Commit:
Fixed frame rate issue in structureFromMotion.py

v2thegreat reviewed (changes requested):
- Write a README.md explaining what this project does*
- Write a TODO.md listing things to do in the future and remove TODO mentions from comments (line 86, 89, 124)**
- Add comments for update (line: 170)
- Break functions down further and following the same guidelines***
- Move code from the line: 175+ into a main-function
- Create a runner/main/getStructureFromMotion/<anything else you find appropriate> function in class: StructureFromMotion that does what this class is designed to do.
- Add unit tests****
- Use common sense, all these things are more guidelines. If you end up choosing not to follow any of these, GIVE A REASON FOR IT. Put that reason either in the Docs or comments, but give it somewhere easy to read. The difference between having someone read your code and saying: "man, you're an idiot" and "man, that's a good point" is how you explain your code.

*: Readme should (briefly) explain the concepts used in this project, give sources for additional reading if needed/can explain the concepts better, such as youtube vids
**: One of the issues with the in-comment TODO's is that they can be easily buried and hard to confirm that we've actually done it before or not, and could potentially waste time. Look at my project: https://github.com/v2thegreat/3D-Printer-Guardian/blob/master/TODO.md for a better idea
***: to quote Clean Code by Robert Cecil Martin, "A function should do one thing and one thing only!". Break the function down into smaller private functions for encapsulation (python doesn't completely support private functions, but you can still enforce the concepts. Look here: https://stackoverflow.com/a/1547163) and readability.
****: This is to make sure that newer code doesn't break the older code, and is better than no tests. We might implement more in the future with running the simulation to make sure that everything still runs

Commit:
Extended distance estimation to support camera rotation

Drew-Wagner committed:
Merge pull request #1 from Drew-Wagner/camera_rotations

Extended distance estimation to support camera rotation

v2thegreat commented on Research and Development Projects/Structure From Motion/structureFromMotion.py:
```diff
@@ -1,221 +1,264 @@
+		:return: float: distance along z axis from camera
+
+		Calculates the z-axis position from the disparity.
+		TODO Or does it? Verify that it is not simply the straight line distance from camera...
```
Remove TODO and move to TODO.md

v2thegreat replied:
Remove TODO comment and move to TODO.md

v2thegreat commented on Research and Development Projects/Structure From Motion/structureFromMotion.py:
```diff
@@ -1,221 +1,264 @@
+		TODO Or does it? Verify that it is not simply the straight line distance from camera...
+		"""
+		# Calculate the vector pointing from the old feature position to the center of the frame (Vanishing point)
+		# TODO Try normalizing with Fast InvSqrt()
```
Remove TODO comment and move to TODO.md

v2thegreat commented on Research and Development Projects/Structure From Motion/structureFromMotion.py:
```diff
@@ -1,221 +1,264 @@
+		"""
+		# Calculate the vector pointing from the old feature position to the center of the frame (Vanishing point)
+		# TODO Try normalizing with Fast InvSqrt()
+		# https://en.wikipedia.org/wiki/Fast_inverse_square_root
```
Place this in README.md with more elaboration if needed

v2thegreat commented on Research and Development Projects/Structure From Motion/structureFromMotion.py:
```diff
@@ -1,221 +1,264 @@
+		disparity = new - old
+		distance = self.calculate_distance_from_disparity(old, disparity)
+
+		# TODO Fix warping of position near edges
```
Remove TODO comment and move to TODO.md

v2thegreat commented on Research and Development Projects/Structure From Motion/structureFromMotion.py:
```diff
@@ -1,221 +1,264 @@
+			for i, (new, old) in enumerate(zip(good_new, good_old)):
+				cam_pos = self.calculate_camera_space_position_of_feature(new, old)
+				if not self.find_in_camera_space:
+					# TODO Camera rotation
```
Remove TODO comment and move to TODO.md

v2thegreat commented on Research and Development Projects/Structure From Motion/structureFromMotion.py:
```diff
@@ -1,221 +1,264 @@
 structureFromMotion = StructureFromMotion(camera_pos=np.array([0, 0, 0]),
-                                          camera_vel=np.array((0, 0, 5.2)),
-                                          camera_space=True)
+										  camera_vel=np.array((0, 0, 0)),
```
Fix Formatting

v2thegreat commented on Research and Development Projects/Structure From Motion/structureFromMotion.py:
```diff
@@ -1,221 +1,264 @@
+										  camera_vel=np.array((0, 0, 0)),
+                                          camera_rot=np.array([0,0,0]),
+                                          camera_rvel=np.array([0,2,0]),
+										  camera_space=True)
```
Fix Formatting

v2thegreat commented on Research and Development Projects/Structure From Motion/structureFromMotion.py:
```diff
@@ -1,221 +1,264 @@
+                                          camera_rot=np.array([0,0,0]),
+                                          camera_rvel=np.array([0,2,0]),
+										  camera_space=True)
 ret, first_frame = cap.read()
```
Please move this into a separate function or put it into
```
if __name__ == "__main__":
    main()
```

This is because if we were to import this into another module, we'd have this code run instead before. 

Please put this into the `main` function and call it into the code block above

v2thegreat commented on Research and Development Projects/Structure From Motion/structureFromMotion.py:
```diff
@@ -1,221 +1,264 @@
+
+	dt = 1 / 20
+	print("Processing video...")
+	while True:
```
Break this down further into smaller functions

v2thegreat commented on Research and Development Projects/Structure From Motion/structureFromMotion.py:
```diff
@@ -1,221 +1,264 @@
+	print("Processing video...")
+	while True:
+		# Get time at beginning of loop for frame rate calculation
+		if dt > 0:
```
Break this down further into smaller functions

v2thegreat commented on Research and Development Projects/Gesture Recognition/Colored Dot Tracking/coloredDotTracking.py:
```diff
@@ -0,0 +1,77 @@
+		return mask
+
+
+red_CDF = ColoredDotFinder("red", Color(0, 127, 127), Color(20, 255, 255))
```
Please move this into a separate function or put it into
```
if __name__ == "__main__":
    main()
```
This is because if we were to import this into another module, we'd have this code run instead before.

Please put this into the `main` function and call it into the code block above

v2thegreat commented on Research and Development Projects/Structure From Motion/structureFromMotion.py:
```diff
@@ -0,0 +1,264 @@
+		self.camera_rot = camera_rot
+		self.camera_rvel = camera_rvel
+
+		# TODO Try with SIFT features
```
Remove TODO comment and move to TODO.md

v2thegreat commented on Research and Development Projects/Structure From Motion/structureFromMotion.py:
```diff
@@ -0,0 +1,264 @@
+		pitch_disparity = y_disparity
+
+		# Weight each axis of the disparity according to camera velocity (dot product)
+		weighted_disparity = x_disparity * self.camera_vel[0] + \
```
Please re-written into a private function

v2thegreat commented on Research and Development Projects/Structure From Motion/structureFromMotion.py:
```diff
@@ -0,0 +1,264 @@
+
+		return np.array([x, y, distance])
+
+	def initialize(self, raw_previous_frame):
```
more detailed names?

v2thegreat commented on Research and Development Projects/Structure From Motion/structureFromMotion.py:
```diff
@@ -0,0 +1,264 @@
+
+		return frame_points
+
+	def update(self):
```
more detailed names?

v2thegreat commented on Research and Development Projects/Structure From Motion/structureFromMotion.py:
```diff
@@ -0,0 +1,264 @@
+			cv2.putText(image, "%.1f" % p[2], (x, y), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 0, 0), 1)
+
+
+video_source = "rotationsample.avi"
```
Please move this into a separate function or put it into
```
if __name__ == "__main__":
    main()
```
This is because if we were to import this into another module, we'd have this code run instead before.

Please put this into the `main` function and call it into the code block above

v2thegreat reviewed (changes requested):
Good work! Nah, _GREAT WORK_!

There are still some small changes that I want you to do to, but other than that, it's perfect!

Commit:
Created README and TODO

Commit:
Added content to README

Commit:
Fixed indentation issue in README

Drew-Wagner committed:
Update TODO.md

Added todos for README

v2thegreat commented on Research and Development Projects/Structure From Motion/README.md:
```diff
@@ -0,0 +1,42 @@
+## Additional Reading
+- https://docs.opencv.org/3.3.1/d7/d8b/tutorial_py_lucas_kanade.html
+- https://ieeexplore.ieee.org/document/6749315
+- https://etd.ohiolink.edu/rws_etd/document/get/dayton1366386933/inline
```
Excellent Work!

v2thegreat commented on Research and Development Projects/Gesture Recognition/Colored Dot Tracking/coloredDotTracking.py:
```diff
@@ -45,8 +45,71 @@ def find_color(self, image, show=False):
+		return self.get_centroid_of_contour(hand_contour)
+
 
-red_CDF = ColoredDotFinder("red", Color(0, 127, 127), Color(20, 255, 255))
```
Still no main block :(

v2thegreat committed:
- Refactored `__init__` for `StructureFromMotion`
  so that initialization of each item is further broken down based on
  where they are used

- Added Comments and Docstrings to all new functions that were created
  for initialization

- Converted `preprocess_frame` to static function

- Minor changes to docstrings for `preprocess_frame`

- Major changes to `calculate_distance_from_disparity`
  Broke down calculate_distance_from_disparity into smaller functions:
  - `_calcVanishingPoint`
  - `_distributeDisparity`
  - `_calcRotationalDisparities`
  - `_calcWeightedDisparity`
  - `_calcDistance`

- Added docstrings for:
  - `_calcVanishingPoint`
  - `_distributeDisparity`
  - `_calcRotationalDisparities`
  - `_calcWeightedDisparity`
  - `_calcDistance`

v2thegreat committed:
Converted preprocess_frame to staticmethod

v2thegreat committed:
Minor Changes

v2thegreat committed:
Merge branch 'master' of https://github.com/v2thegreat/uav-swarm-project

v2thegreat committed:
Drew's Mods

Drew-Wagner committed:
Merge pull request #2 from v2thegreat/master

Refactored parts of functions
